spark_log_parser/loaders/__init__.py

```diff
@@ -61,58 +71,110 @@ def __iter__(self):
+                self._buffer_memview = memoryview(self._buffer).cast("B").toreadonly()
+
+            # If we have enough data in the buffer to satisfy this read, return that from the buffer and adjust
+            #  our read position accordingly
+            if self._num_bytes_left_in_buffer() >= size:
+                start_pos = self._buffer_position
+                self._buffer_position += size
+                return self._buffer_memview[start_pos:self._buffer_position]
+            else:
+                # Otherwise, we don't have enough data, which means that this is the last read for this iterator.
+                #  So now just return a copy of the data remaining in the buffer and mark this as consumed to prevent
+                #  future reads
+                content = self._buffer_memview[self._buffer_position:]
+                self._consumed = True
+                return content
+
+    _LINE_CHUNK_SIZE = 1024 * 1024
 
     def iter_lines(self) -> Iterator[bytes]:
         """
         Returns a stream of lines from self._chunks.
         """
-        for chunk in self._chunks:
-            self._add_chunk_to_size(chunk)
-            if self._trailing_data:
-                chunk = self._trailing_data + chunk
-
-            lines = chunk.splitlines()
+        trailing_data = bytearray()
+        # We use a relatively large chunk_size here to minimize the number of times we have to enter this loop/copy data
+        while chunk := self.read(self._LINE_CHUNK_SIZE):
+            if trailing_data:
+                # Instead of appending via e.g. chunk = trailing_data + chunk, which causes a ton of data copying,
+                #  we can just write this chunk into our current line buffer. This is extremely helpful when we
+                #  encounter large files where all the data is actually on one line (i.e. a 500MB JSON file that was
+                #  written without any line breaks). This saves us 10s of seconds when processing these types of files
+                # This is safe to do here because trailing_data is never referenced outside of this function, and we
+                #  only ever return trailing_data when we are done reading in the whole file
+                trailing_data.extend(chunk)
+                chunk = trailing_data
+                trailing_data = bytearray()
+
+            # In the case where we have no trailing_data, then our `chunk` may be a memoryview returned from
+            #  self.read(), which does not support splitlines(), so we need to copy this into a bytearray first
+            bs = chunk if isinstance(chunk, bytearray) else bytearray(chunk)
+
+            # Don't check for `\r\n` because that will be caught by checking for the individual characters
+            # Doing this similarly helps keep parsing fast for large single-line files, while having minimal performance
+            #  impact on files that actually have more than 1 line
+            if b'\n' in bs or b'\r' in bs:
```

Comment:
Couldn't we just look for `b'\n'`?

Reply:
Yeah, it's a good question, and I think it depends how "correct"/robust we'd like to be. `splitlines()` on `bytes`/`bytearray` considers all three of `\n`, `\r`, and `\r\n` as valid newline sequences - https://docs.python.org/3/library/stdtypes.html#bytes.splitlines. So I included `\r` here in order to be compatible with that. In practice, though, I agree with you - I would be pretty surprised if we received a file using lone `\r` characters for newlines... Python has to be a lot more "proper" in this respect, but we may not need to be.